File: skills/code-architect/references/code-examples/php/Prototype.php
<?php
// Prototype — copy existing objects via a clone interface without coupling to classes.
declare(strict_types=1);

interface Cloneable
{
    public function duplicate(): static;
}

final class Tag
{
    public function __construct(public string $label) {}
}

final class Shape implements Cloneable
{
    /** @param Tag[] $tags */
    public function __construct(
        public string $kind,
        public int $size,
        public array $tags = [],
    ) {}

    public function duplicate(): static
    {
        $copy = clone $this;
        // Deep-copy mutable members so the clone is independent.
        $copy->tags = array_map(static fn (Tag $t) => clone $t, $this->tags);
        return $copy;
    }

    public function describe(): string
    {
        $labels = array_map(static fn (Tag $t) => $t->label, $this->tags);
        return "{$this->kind}({$this->size})[" . implode(',', $labels) . ']';
    }
}

// Demo
$original = new Shape('circle', 10, [new Tag('a'), new Tag('b')]);
$copy = $original->duplicate();
$copy->size = 20;
$copy->tags[0]->label = 'z';

echo $original->describe() . PHP_EOL; // unchanged
echo $copy->describe() . PHP_EOL;

File: skills/code-architect/references/code-examples/go/prototype.go
// Prototype: produce new objects by cloning an existing instance instead
// of constructing them from scratch.
package main

import "fmt"

// Cloneable is anything that can copy itself.
type Cloneable interface {
	Clone() Cloneable
	Describe() string
}

// Node holds value state plus a slice to show deep copying.
type Node struct {
	label string
	tags  []string
}

func (n *Node) Clone() Cloneable {
	tags := make([]string, len(n.tags))
	copy(tags, n.tags)
	return &Node{label: n.label, tags: tags}
}

func (n *Node) Describe() string {
	return fmt.Sprintf("%s%v", n.label, n.tags)
}

func main() {
	original := &Node{label: "origin", tags: []string{"x", "y"}}
	copy := original.Clone().(*Node)
	copy.label = "copy"
	copy.tags = append(copy.tags, "z")

	fmt.Println("original:", original.Describe())
	fmt.Println("clone:   ", copy.Describe())
}
